internal/app/machined/pkg/controllers/k8s/kubelet_service.go

```diff
@@ -297,3 +324,86 @@ func updateKubeconfig(newEndpoint *url.URL) error {
 
 	return clientcmd.WriteToFile(*config, constants.KubeletKubeconfig)
 }
+
+// refreshKubeletCerts checks if the existing kubelet certificates match the node hostname.
+// If they don't match, it clears the certificate directory and the removes kubelet's kubeconfig so that
+// they can be regenerated next time kubelet is started.
+func (ctrl *KubeletServiceController) refreshKubeletCerts(hostname string) error {
+	cert, err := ctrl.readKubeletCertificate()
+	if err != nil {
+		return err
+	}
+
+	if cert == nil {
+		return nil
+	}
+
+	valid := slices.Contains(cert.DNSNames, func(name string) bool {
+		return name == hostname
+	})
+
+	if valid {
+		// certificate looks good, no need to refresh
+		return nil
+	}
+
+	dir, err := os.ReadDir(constants.KubeletPKIDir)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// clear the pki directory
+	for _, entry := range dir {
+		err = os.RemoveAll(filepath.Join(constants.KubeletPKIDir, entry.Name()))
+		if err != nil {
+			return err
+		}
+	}
+
+	// clear the kubelet kubeconfig
+	err = os.Remove(constants.KubeletKubeconfig)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	return err
+}
+
+func (ctrl *KubeletServiceController) readKubeletCertificate() (*x509.Certificate, error) {
+	raw, err := os.ReadFile(filepath.Join(constants.KubeletPKIDir, "kubelet.crt"))
+	if errors.Is(err, os.ErrNotExist) {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	for {
```

Comment:
is this because the file contains multiple certs?

Reply:
yep, one is CA and other is the server one.